feat(client-debugger): Update audience view to use Fluent UI Elements

jikim-msft opened the pull request:
#### Description
https://dev.azure.com/fluidframework/internal/_workitems/edit/3865/

This PR applies `@fluentui/react` to `fluid-devtool`'s Audience State and Audience History panel. 

https://react.fluentui.dev/?path=/docs/components-table--default

With the introduction of `Table`, clients are now able to see data in more details organized in columns instead of toggling the section.

#### TODO
- ~~The current `fluentUI-table` component is verbose as it requires `Table`, `TableHeader`, `TableRow` and etc. Creating a separate component for Table should be considered.~~
- ~~Demo in `fluentUI-table` renders icon per each data, feedback on whether to emulate the demo or to render icon solely on column would be appreciated.~~
- `audienceStateColumns` and `audienceHistoryColumns` are declared within `AudienceView` as constant value, should consider if there would be a better approach. 
- Add `tooltips` explaining columns in more detail (would need to clarify ambiguity in difference between `userId` vs. `clientId`) 

Before 
<img width="1086" alt="image" src="https://user-images.githubusercontent.com/111468570/230307093-0a4719da-ba12-4a5e-aba0-a5842ab58ea8.png">


After
<img width="961" alt="image" src="https://user-images.githubusercontent.com/111468570/230466904-5b15ccf1-f706-48e8-b4e3-ea06105cf4ab.png">

jikim-msft committed:
apply fluentUI to audience lit

jikim-msft committed:
intall fluentui/icon

jikim-msft committed:
Merge branch 'fluent-audience-list' of https://github.com/jikim-msft/FluidFramework into fluent-audience-list

jikim-msft committed:
Merge branch 'main' into fluent-audience-list

jikim-msft committed:
Merge branch 'main' into render-audience-list

jikim-msft committed:
Merge branch 'fluent-audience-list' of https://github.com/jikim-msft/FluidFramework into fluent-audience-list

rohandubal commented:
Few thoughts:
- Given the columns are different for the 2 scenarios "current audience" and "past audience" we should have 2 separate tables with a partition/divider in between. 
- For current audience, can we also include the timestamp when the member joined the session (if available)
- Can we highlight the row for the current user?
- For audience that have left the session, we can also apply one of the light red colors from red palette to indicate they have left the session 
- I would lean towards including the icon in column header and not for each row

jikim-msft committed:
add icon and separate components

jikim-msft committed:
add/remove color visualization

jikim-msft committed:
identify current user

jikim-msft committed:
prettier fix

jikim-msft commented:
> Few thoughts:
> 
> * Given the columns are different for the 2 scenarios "current audience" and "past audience" we should have 2 separate tables with a partition/divider in between.
> * For current audience, can we also include the timestamp when the member joined the session (if available)
> * Can we highlight the row for the current user?
> * For audience that have left the session, we can also apply one of the light red colors from red palette to indicate they have left the session
> * I would lean towards including the icon in column header and not for each row


1. `AudienceState` and `AudienceHistory` are now divided with `fluentUI`'s `Divider`

2. Would have to look inside, if I can access the timestamp from `audienceData.audienceState`. `audienceData.audienceHistory` accesses the timestamp by `timestamp` in `readonly AudienceChangeLogEntry`. Definitely worth looking into. 

https://github.com/microsoft/FluidFramework/blob/f453d5688ae64af14c2731bea09608283ffd170c/packages/tools/client-debugger/client-debugger-view/src/components/AudienceView.tsx#L162

3. Now visualizes by the background. 

https://github.com/microsoft/FluidFramework/blob/f453d5688ae64af14c2731bea09608283ffd170c/packages/tools/client-debugger/client-debugger-view/src/components/AudienceStateTable.tsx#L55

4. Also applies the similar logic from above to distinguish the client status by background 

https://github.com/microsoft/FluidFramework/blob/f453d5688ae64af14c2731bea09608283ffd170c/packages/tools/client-debugger/client-debugger-view/src/components/AudienceHistoryTable.tsx#L51

5. Icon added in the column level for both `AudieceState` and `AudienceHistory`

https://github.com/microsoft/FluidFramework/blob/f453d5688ae64af14c2731bea09608283ffd170c/packages/tools/client-debugger/client-debugger-view/src/components/AudienceStateTable.tsx#L41

https://github.com/microsoft/FluidFramework/blob/f453d5688ae64af14c2731bea09608283ffd170c/packages/tools/client-debugger/client-debugger-view/src/components/AudienceHistoryTable.tsx#L39

alexvy86 commented on packages/tools/client-debugger/client-debugger-view/src/components/AudienceHistoryTable.tsx:
```diff
@@ -0,0 +1,61 @@
+	const { audienceHistoryItems } = props;
+
+	// Columns for rendering audience history
+	const audienceHistoryColumns = [
+		{ columnKey: "clientId", label: "ClientId" },
+		{ columnKey: "time", label: "Time" },
+	];
```
Nit: maybe we want to add a third column for "Event" with values Join/Leave? For users who are seeing this information for the first time, the color coding might not be enough to fully explain what each row is about. cc @rohandubal ?

jikim-msft replied:
Interesting. I agree that color could not be enough for users using the devtools for the first time. I would like to have more thoughts on this.

Josmithr replied:
Also for accessibility purposes, we can't depend strictly on colors. We should definitely re-add the icons somewhere. I personally like the "event" column idea.

Josmithr replied:
And actually, I almost wonder if the coloring is too bright / noisy. Maybe **just** icons would be better?

jikim-msft replied:
`event` column added to note if the audience was `added` or `removed`

Josmithr replied:
Can we move the event column to be the first column? I think that's the most important contextual info we're presenting in that view.

alexvy86 commented on packages/tools/client-debugger/client-debugger-view/src/components/AudienceHistoryTable.tsx:
```diff
@@ -0,0 +1,61 @@
+					<TableRow
+						key={itemIndex}
+						style={{
+							backgroundColor: item.changeKind === "added" ? "#90ee90" : "#FF7377",
```
This should be using the `tokens` object Rohan showed us for the colors, right?

jikim-msft replied:
Resolved.

alexvy86 commented on packages/tools/client-debugger/client-debugger-view/src/components/AudienceStateTable.tsx:
```diff
@@ -0,0 +1,75 @@
+							backgroundColor:
+								item.myClientConnection !== undefined &&
+								item.myClientConnection.user.id === item.userId
+									? "#add8e6"
```
Same here.

jikim-msft replied:
Resolved.

alexvy86 commented on packages/tools/client-debugger/client-debugger-view/src/components/AudienceStateTable.tsx:
```diff
@@ -0,0 +1,75 @@
+						<TableCell>{item.mode}</TableCell>
+						<TableCell>
+							{item.scopes.map((scope, scopeIndex) => (
+								<div key={scopeIndex}>{scope}</div>
```
Nit: I'd just use text and `<br />` elements, a `div` for each entry feels slightly overkill. Maybe something like `item.scopes.join("<br />")`.

jikim-msft replied:
Is replacing `<div/>` using `<pre/>` tag as `<br/>` complains an error.

alexvy86 commented on packages/tools/client-debugger/client-debugger-view/src/components/AudienceView.tsx:
```diff
@@ -87,158 +86,92 @@ export function AudienceView(props: AudienceViewProps): React.ReactElement {
+	myClientConnection: IClient | undefined,
+): FilteredAudienceStateData[] {
+	return audienceStateData.map((entry) => {
+		const clientId = entry.clientId;
+		const userId = entry.client.user.id;
+		const mode = entry.client.mode;
+		const scopes = entry.client.scopes;
+
+		return {
+			clientId,
+			userId,
+			mode,
+			scopes,
+			myClientConnection,
+		};
```
Nit; since those variables are only used once, "merging" the assignments into the object construction saves us some lines.
```suggestion
		return {
			clientId: entry.clientId,
			userId: entry.client.user.id,
			mode: entry.client.mode,
			scopes: entry.client.scopes,
			myClientConnection,
		};
```

jikim-msft replied:
Resolved.

alexvy86 commented on packages/tools/client-debugger/client-debugger-view/src/components/AudienceView.tsx:
```diff
@@ -87,158 +86,92 @@ export function AudienceView(props: AudienceViewProps): React.ReactElement {
-			root: {
-				padding: 10,
-			},
+		return {
+			clientId,
+			time,
+			changeKind,
 		};
```
Same here (will need prettier run on it):
```suggestion
		return {
			clientId: entry.clientId,
			time: wasChangeToday
			? changeTimeStamp.toTimeString()
			: changeTimeStamp.toDateString(),
			changeKind: entry.changeKind,
		};
```

jikim-msft replied:
Resolved.

alexvy86 commented on packages/tools/client-debugger/client-debugger-view/src/components/AudienceView.tsx:
```diff
@@ -87,158 +86,92 @@ export function AudienceView(props: AudienceViewProps): React.ReactElement {
-	}
-
-	return <Stack>{memberViews}</Stack>;
+function AudienceStateDataFilter(
+	audienceStateData: AudienceClientMetadata[],
+	myClientConnection: IClient | undefined,
+): FilteredAudienceStateData[] {
```
I'd rename this, the `AudienceHistoryDataFilter`, and the two related interfaces, since they're not doing any filtering.

Josmithr replied:
Also functions should be in `camelCase`

jikim-msft replied:
I believe the two helper functions serve the general purpose of "filtering" unnecessary data from the `audienceData.audienceState` and `audienceData.audienceHistory`.

At first, the two JavaScript objects contain unnecessary metadata which is irrelevant to the information rendered for the `AudienceView` so I think the `fooDataFilter` is pertinent for the naming?

alexvy86 replied:
I'd call that transformation more than filtering though; we're changing the shape of the data to something more suited to the task/component at hand, but we're not skipping any entries from the original list. "Filter" has the connotation of choosing which entries in a list or table should or shouldn't be considered for display or some next step (no better example than JS arrays' `filter()` method 😄). I don't know that these particular tables will ever need filtering like that, but I'd still recommend we don't overload the term.

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/components/AudienceHistoryTable.tsx:
```diff
@@ -0,0 +1,61 @@
+ * Input for {@link AudienceHistoryTable}
+ */
+export interface AudienceHistoryTableProps {
+	audienceHistoryItems: FilteredAudienceHistoryData[];
```
Let's document this as well.

jikim-msft replied:
Resolved.

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/components/AudienceStateTable.tsx:
```diff
@@ -0,0 +1,75 @@
+ * Input for {@link AudienceStateTable}
+ */
+export interface AudienceStateTableProps {
+	audienceStateItems: FilteredAudienceStateData[];
```
Let's document this too.

jikim-msft replied:
Resolved.

Josmithr commented:
Feedback on the visuals:

1. The bright blue background color differentiating the "me" client doesn't work well with the white text. It's pretty hard to read.
    - Also, we should add some sort of icon or something to distinguish the "me" client in addition to (or instead of) the background color, for accessibility.
2. The icons we're using for `ClientId` and `UserId` have circles behind them, but the other icons don't. We should try to be consistent here (probably no circles?).
3. We should write out "Client ID" instead of "ClientId" and "User ID" instead of "UserId".

jikim-msft committed:
add visual effect and me sign

jikim-msft commented:
> Feedback on the visuals:
> 
> 1. The bright blue background color differentiating the "me" client doesn't work well with the white text. It's pretty hard to read. Darker colors would probably be better if we want to keep the background colors. See [Wayne's PR](https://github.com/microsoft/FluidFramework/pull/14961) for reference.
>    
>    * Also, we should add some sort of icon or something to distinguish the "me" client in addition to (or instead of) the background color, for accessibility.
> 2. The icons we're using for `ClientId` and `UserId` have circles behind them, but the other icons don't. We should try to be consistent here (probably no circles?).
> 3. We should write out "Client ID" instead of "ClientId" and "User ID" instead of "UserId".

1. I changed the background to darker color to make it compatible with white text. I thought about adding an icon to distinguish "me", but after some attempts decided to go with `"me"`. I personally think embedding an icon on a single row while other don't is not consistent. 
2. I changed the icon style to conform to other icons. 
3. Resolved.

jikim-msft committed:
Merge branch 'microsoft:main' into fluent-audience-list

jikim-msft committed:
Merge branch 'microsoft:main' into fluent-audience-list

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/components/AudienceHistoryTable.tsx:
```diff
@@ -0,0 +1,76 @@
+}
+
+/**
+ * Render audience history in {@link AudienceView}
```
Generally it's best to avoid documenting components in terms of what consumes them, since that can change. Let's instead document this in terms of what it displays.

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/components/AudienceHistoryTable.tsx:
```diff
@@ -0,0 +1,76 @@
+ */
+export interface AudienceHistoryTableProps {
+	/**
+	 * Filtered audience data from {@link audienceHistoryDataFilter}
```
I would recommend documenting this in terms of what it means to the component consuming it, rather than how it is generated. E.g. This represents audience data filtered to the attributes that will be displayed in the table,

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/components/AudienceView.tsx:
```diff
@@ -89,158 +88,81 @@ export function AudienceView(props: AudienceViewProps): React.ReactElement {
-	}
-
-	return <Stack>{memberViews}</Stack>;
+function audienceStateDataFilter(
```
I generally recommend naming functions to start with a verb, so it's clear that it represents an action of some kind. E.g. "filterAudienceStateData".

That said, since this is only used in 1 place, we could probably just inline the call to `.map` into the component too.

jikim-msft replied:
Based on the earlier conversation with @alexvy86, I changed the name to `transformAudienceStateData`.

Josmithr commented on packages/tools/client-debugger/client-debugger-view/src/components/AudienceView.tsx:
```diff
@@ -89,158 +88,81 @@ export function AudienceView(props: AudienceViewProps): React.ReactElement {
-function HistoryView(props: HistoryViewProps): React.ReactElement {
-	const { history } = props;
-
+function audienceHistoryDataFilter(
```
I recommend "filterAudienceHistoryData" instead.

Josmithr reviewed (approved):
I left a few more recommendations, but I think the shape of this looks great! Thanks for doing this!

jikim-msft committed:
Merge branch 'microsoft:main' into fluent-audience-list

jikim-msft committed:
change to inline / change documentation / change column

jikim-msft committed:
run prettier
